pkg/registry/registry_to_model.go

```diff
@@ -0,0 +1,241 @@
+package registry
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/operator-framework/operator-registry/pkg/model"
+	"github.com/operator-framework/operator-registry/pkg/property"
+)
+
+func CreateModelFromBundle(b *Bundle) (*model.Model, error) {
+	desc, err := b.csv.GetDescription()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get description from bundle CSV:%s", err)
+	}
+	i, err := b.csv.GetIcons()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get icon from bundle CSV:%s", err)
+	}
+	mIcon := &model.Icon{
+		// looks like the CSV icon is a slice of objects https://github.com/operator-framework/api/blob/master/pkg/operators/v1alpha1/clusterserviceversion_types.go#L283
+		// Any better idea here than just taking the first element of the slice and working with that? Since I'm assuming in most cases users only pass in one element (icon)
+		// in the CSV anyway
+		MediaType: i[0].MediaType,
+		Data:      []byte(i[0].Base64data),
+	}
+	pkg := &model.Package{
+		Name:        b.Annotations.PackageName,
+		Description: desc,
+		Icon:        mIcon,
+	}
+	mb, err := registryBundleToModelBundle(b)
+	mb.Package = pkg
+	if err != nil {
+		return nil, err
+	}
+	mbs := make(map[string]*model.Bundle)
+	mbs[b.csv.Name] = mb
+	annotationsChannels := strings.Split(b.Annotations.Channels, ",")
+	channels := make(map[string]*model.Channel)
+	// This is based on the assumption that operators.operatorframework.io.bundle.channels.v1 in
+	// the annotations.yaml file only accepts channels as comma seperated list of strings. Is this
+	// correct?
+	for i, ch := range annotationsChannels {
+		modelChannel := model.Channel{
+			Name:    ch,
+			Package: pkg,
+			Bundles: mbs,
+		}
+		channels[ch] = &modelChannel
+		if ch == b.Annotations.DefaultChannelName {
+			pkg.DefaultChannel = &modelChannel
+		}
+		if i+1 == len(annotationsChannels) && pkg.DefaultChannel == nil {
+			pkg.DefaultChannel = &modelChannel
+		}
+	}
+	pkg.Channels = channels
+	pkgs := model.Model{}
+	pkgs[b.Annotations.PackageName] = pkg
+	return &pkgs, nil
+}
+
+func registryBundleToModelBundle(b *Bundle) (*model.Bundle, error) {
+	bundleProps, err := convertRegistryBundleToModelProperties(b)
+	if err != nil {
+		return nil, fmt.Errorf("convert properties: %v", err)
+	}
+
+	csv, err := b.ClusterServiceVersion()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get CVS for bundle: %s", err)
+	}
+	replaces, err := csv.GetReplaces()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get Replaces from CSV for bundle: %s", err)
+	}
+	skips, err := csv.GetSkips()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get Skips from CSV for bundle: %s", err)
+	}
+	relatedImages, err := convertToModelRelatedImages(csv)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get Related images from bundle: %v", err)
+	}
+
+	return &model.Bundle{
+		Name:          csv.Name,
+		Image:         b.BundleImage,
+		Replaces:      replaces,
+		Skips:         skips,
```

Comment:
I kind of wonder if we should specifically ignore `replaces` (and maybe `skips`) values from the CSV when adding bundles.

If we _don't_ ignore them, we're effectively making `opm add` build upgrade graphs in `replaces` mode, aren't we?

If we _do_ ignore them, a follow-on tool _must_ be used to insert the added bundles into the upgrade graph.

Reply:
I don't think adding `replaces` from the CSV here is essentially "building" the graph. It's just adding the metadata, i.e expressing the *intent* to build the graph. I don't think it's possible to say you must use another tool always to effectively make your configs make sense.

Reply:
But `replaces` and `skips` define the graph. So if they're both in there, we effectively have built the graph, right?

Reply:
`replaces` and `skips` define the update graph in `replaces` mode for sure. We can't simply ignore them because users include them to build the update graph unless they are using `semver` and `skippatch` mode.

Reply:
"We" as in opm ins't building the graph, opm is simply translating the intent of the user of opm (communicated via the bundle manifests) onto the configs. Essentially the user is building the graph, we're merely transplanting the very thing the user communicated in the bundle in a very non-opinionated manner onto the configs.

Reply:
I agreed. We are moving away from the `mode` style and this should be documented somewhere.